scripts/run_portserver.py

```diff
@@ -0,0 +1,531 @@
+          available ports with no active process using them is getting
+          low.
+    """
+
+    def __init__(self):
+        self._port_queue = collections.deque()
+        self.ports_checked_for_last_request = 0
+
+    def num_ports(self):
+        """Get the number of ports in the pool.
+
+        Returns:
+            int. The number of ports in the pool.
+        """
+        return len(self._port_queue)
+
+    def get_port_for_process(self, pid):
+        """Allocates port.
+
+        Args:
+            pid: int. ID for process to allocate port to.
+
+        Returns:
+            int. Allocated port or 0 if none could be allocated.
+        """
+        if not self._port_queue:
+            raise RuntimeError('No ports being managed.')
+
+        # Avoid an infinite loop if all ports are currently assigned.
+        check_count = 0
+        max_ports_to_test = len(self._port_queue)
+        while check_count < max_ports_to_test:
+            # Get the next candidate port and move it to the back of the queue.
+            candidate = self._port_queue.pop()
+            self._port_queue.appendleft(candidate)
+            check_count += 1
+            if (candidate.start_time == 0
+                    or candidate.start_time
+                    != _get_process_start_time(candidate.pid)):
+                if _is_port_free(candidate.port):
+                    candidate.pid = pid
+                    candidate.start_time = _get_process_start_time(pid)
+                    if not candidate.start_time:
+                        logging.info('Can\'t read start time for pid %d.', pid)
+                    self.ports_checked_for_last_request = check_count
+                    return candidate.port
+                else:
+                    logging.info(
+                        'Port %d unexpectedly in use, last owning pid %d.',
+                        candidate.port, candidate.pid)
+
+        logging.info('All ports in use.')
+        self.ports_checked_for_last_request = check_count
+        return 0
+
+    def add_port_to_free_pool(self, port):
+        """Add a new port to the free pool for allocation.
+
+        Args:
+            port: int. The port number to add to the pool.
+        """
+        if port < 1 or port > 65535:
+            raise ValueError(
+                'Port must be in the [1, 65535] range, not %d.' % port)
+        port_info = _PortInfo(port=port)
+        self._port_queue.append(port_info)
+
+
+class _PortServerRequestHandler(python_utils.OBJECT):
+    """A class to handle port allocation and status requests.
+
+    Allocates ports to process ids via the dead simple port server protocol
+    when the handle_port_request asyncio.coroutine handler has been registered.
+    Statistics can be logged using the dump_stats method.
+    """
+
+    def __init__(self, ports_to_serve):
+        """Initialize a new port server.
+
+        Args:
+            ports_to_serve: list. A sequence of unique port numbers to
```

Comment:
list(int). (I think? But the point is -- please check, for all lists, that you specify the types of their elements, too.)

Reply:
done in b6a7f24